### DISCUSSION TITLE
Automatically handle basic math functions in hybrid properties

### DISCUSSION THREAD

**Discussion by ludaavics** [sqlalchemy/sqlalchemy#8557] [Usage Questions] [answered]
### Describe the use case

When using basic math functions in hybrid properties, we have to write the instance and class expressions separately. Code would be easier to maintain if we could forego the latter for the sql-defined math operations
https://www.postgresql.org/docs/current/functions-math.html

### Databases / Backends / Drivers targeted

Postgresql with pscyop2 and asyncpg

### Example Use

```python
class Event(Base):
    __tablename__ = 'events'
    id = Column(Integer, primary_key=True)
    probability= Column(Float)

    @hybrid_property
    def log_probability(self):
        return math.log(self.probability)

    # can we make it so that we don't need to write the below?
    @log_probability.expression
    def log_probability(cls):
        return func.ln(cls.probability)
```

### Additional context

The example above is trivial but this would be useful on more complex equations, especially when there are `max` or `min` involved.
Would also make it easier to test, since right now you have to test output on the instance and the query separately

**Comment by zzzeek** [comment, marked as answer]
is the idea that there would be some lookup table of "math.log" to "func.ln" ?    I dont see how that's possible without the decorator looking at the bytecode of the log_probability function - that function is just a Python callable that is otherwise opaque to the decorator.   Additionally the SQLAlchemy project would never be able to handle mapping every possible math function on every database, this is something people should implement on their own as needed.

I would address this use case by just building pre-packaged hybrids for the functions you need.  hybrids are basically mostly like [Python property objects](https://docs.python.org/3/library/functions.html?highlight=property#property).  an easy way to get this is:

```py
import math
import operator

from sqlalchemy import Column
from sqlalchemy import Float
from sqlalchemy import func
from sqlalchemy import Integer
from sqlalchemy.ext.hybrid import hybrid_property
from sqlalchemy.orm import declarative_base

Base = declarative_base()


class log(hybrid_property):
    def __init__(self, attrname):
        getter = operator.attrgetter(attrname)
        super().__init__(
            fget=lambda obj: math.log(getter(obj)),
            expr=lambda cls: func.ln(getter(cls)),
        )


class Event(Base):
    __tablename__ = "events"
    id = Column(Integer, primary_key=True)
    probability = Column(Float)

    log_probability = log("probability")


print(Event(probability=.10).log_probability)
print(Event.log_probability + 5)
```

that would also make a good extension package or new sqlalchemy-utils feature.

**Comment by janpeterka** [comment]
I have somewhat similar question, but on other case. I was surprised that my `hybrid_property` is not working correctly:
```python
    @hybrid_property
    def is_current(self) -> bool:
        return self.started_in_past and self.ends_in_future
```
Both `started_in_past` and `ends_in_future` are also hybrid properties, comparing datetimes.  
Using this in filter gave me wrong results, I found out only second condition is used in created SQL query. 
I had to add this:
```python
    @is_current.expression
    def is_current(cls):
        from sqlalchemy import and_
        return and_(cls.started_in_past, cls.ends_in_future)
```

This is quite unintuitive, and also as it's "working", it's pretty difficult to find out it's not giving right results.

Edit:  
Ok, I found out from [other discussion](https://github.com/sqlalchemy/sqlalchemy/discussions/5904) that one solution might be using `&` instead of `and`. In that case, it might be quite easy to have this converted automaticaly by SQLAlchemy?

**Reply by zzzeek** [reply]
this is expected.  the Python "and" operator is not supported at a SQL expression construct, and the main thing to understand about "hybrids" are the way a single Python expression is repurposed in two different contexts; the instance level, against an object with state, and the class level, against a class that will generate SQL expressions.    The section https://docs.sqlalchemy.org/en/14/orm/extensions/hybrid.html#defining-expression-behavior-distinct-from-attribute-behavior explains why you might need to use `@expression` (which is perfectly normal and usually convenient).

**Reply by CaselIT** [reply]
> Edit: Ok, I found out from [other discussion](https://github.com/sqlalchemy/sqlalchemy/discussions/5904) that one solution might be using `&` instead of `and`. In that case, it might be quite easy to have this converted automaticaly by SQLAlchemy?

No, since you cannot override what `and` (`or`, `not`, `is` too) does in python. So a statement like `left and right` always ANDs the truth value of left and right. There is nothing SQLAlcehmy can do here to change that behaviour

**Reply by lelit** [reply]
Am I wrong interpreting that @janpeterka said as "if I write `exprA & exprB`, could SA interpret it as `exprA and exprB` on the Python side?

**Reply by CaselIT** [reply]
if that's the case then it's already like this, so there is no need to add additional logic

**Reply by janpeterka** [reply]
> Am I wrong interpreting that @janpeterka said as "if I write `exprA & exprB`, could SA interpret it as `exprA and exprB` on the Python side?

I meant it the other way - if I write `and` in hybrid_property, convert it automatically to sql way, not having to create special expression "just" to rearrange my conditions and put them inside `and_` - as I did in my code.
But honestly, I don't understand how `hybrid_property` (and SQLAlchemy) works internally. I just thought this would be expected behaviour, and its weird that this `hybrid_property` IS accepted, but the query generated from it is not what one would expect (getting something like `SELECT .... WHERE ends_at > now`, expecting `WHERE ends_at > now AND begins_at < now` (now really being param given to query, just simplified here).

**Reply by CaselIT** [reply]
@janpeterka if you have ideas on how to detect it please suggest. from what I know of python there is no way of doing it

What you have written from python point of view is is the same as writing
```py
def x():
  return 'foo' and 'bar'
```
that returns `'bar'`.
There is no way of knowing that `x()` had an `and` and was not just `return 'bar'`
